append_lifecycle.go

```diff
@@ -54,7 +54,15 @@ type AddFn func(ctx context.Context, entry *Entry) IndexFuture
 //
 // Implementations of this func are likely to be "futures", or a promise to return this data at
 // some point in the future, and as such will block when called if the data isn't yet available.
-type IndexFuture func() (uint64, error)
+type IndexFuture func() (Index, error)
+
+// Index represents a durably assigned index for some entry.
+type Index struct {
+	// Index is the location in the log to which a particular entry has been assigned.
+	Index uint64
+	// IsDup is true if Index represents a previously assignment for the same entry.
```

Comment:
"previously assignment" needs some rewriting. Could also do with some docs explaining that this will only be returned if AntiSpam is enabled etc. I still have some internal conflict about how dupe and anti-spam mean the same thing, but if you only know one term the other is hard to discover.